File: compiler/nodes/patterns/string.go
package patterns

import (
	"strconv"

	"wipple/codegen"
	"wipple/database"
	"wipple/nodes/types"
	"wipple/syntax"
	"wipple/typecheck"
	"wipple/visit"
)

type StringPatternNode struct {
	Value string
	Facts *database.Facts

	matching database.Node
}

func (node *StringPatternNode) GetFacts() *database.Facts {
	return node.Facts
}

func ParseStringPattern(parser *syntax.Parser) (*StringPatternNode, *syntax.Error) {
	span := parser.Spanned()

	value, err := parser.Token("String")
	if err != nil {
		return nil, err
	}

	return &StringPatternNode{
		Value: value,
		Facts: database.NewFacts(span()),
	}, nil
}

func (node *StringPatternNode) Visit(visitor *visit.Visitor) {
	visitPattern(visitor, node)
	node.matching = visitor.CurrentMatch.Node

	stringType := &types.NamedTypeNode{
		Name:       "String",
		Parameters: []database.Node{},
		Facts:      database.NewFacts(database.GetSpanFact(node)),
	}

	visitor.Visit(stringType)

	visitor.Constraint(typecheck.NewGroupConstraint(node, stringType))
}

func (node *StringPatternNode) Codegen(c *codegen.Codegen) error {
	span := database.GetSpanFact(node)

	c.WriteString(span, " && (")
	c.WriteNode(span, node.matching)
	c.WriteString(span, "=== ")
	c.WriteString(span, strconv.Quote(node.Value))
	c.WriteString(span, ")")

	return nil
}

File: compiler/src/nodes/patterns/string.rs
use crate::{
    codegen::{Codegen, CodegenCtx, CodegenResult, ir},
    database::{HiddenNode, Node, NodeRef},
    nodes::{Matching, NamedTypeNode, visit_pattern},
    syntax::{ParseError, Parser, TokenKind},
    typecheck::GroupConstraint,
    visit::{MatchPathSegment, Visit, Visitor},
};

#[derive(Debug)]
pub struct StringPatternNode {
    pub value: String,
}

impl Node for StringPatternNode {}

pub fn parse_string_pattern(parser: &mut Parser<'_>) -> Result<StringPatternNode, ParseError> {
    let value = parser.token(TokenKind::String)?;

    Ok(StringPatternNode { value })
}

impl Visit for StringPatternNode {
    fn visit(&self, node: &NodeRef, visitor: &mut Visitor<'_>) {
        visit_pattern(node, visitor, Some(MatchPathSegment::NoMatch));

        let span = visitor.span(node);
        let string_type = visitor.node(
            span,
            HiddenNode::new(NamedTypeNode {
                name: String::from("String"),
                parameters: Vec::new(),
            }),
        );

        visitor.visit(&string_type);
        visitor.constraint(GroupConstraint::new(node.clone(), string_type));
    }
}

impl Codegen for StringPatternNode {
    fn codegen(&self, node: &NodeRef, ctx: &mut CodegenCtx<'_>) -> CodegenResult {
        let Matching(matching) = ctx
            .get(node)
            .ok_or_else(|| anyhow::format_err!("unresolved"))?;

        ctx.condition(ir::Condition::EqualToString {
            input: matching,
            value: self.value.clone(),
        });

        Ok(())
    }
}
